src/calculations.ts

```diff
@@ -0,0 +1,159 @@
+import { isNumber } from "./utilities";
+
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Neil Enns. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+function celsiusToFahrenheit(celsius: number): number {
+  return (celsius * 9) / 5 + 32;
+}
+
+function fahrenheitToCelsius(fahrenheit: number): number {
+  return ((fahrenheit - 32) * 5) / 9;
+}
+
+/**
+ * Converts solar radiation to solar radiation in lux
+ * @param radiation The solar radiation
+ * @returns The solar radiation in lux
+ */
+export function calculateSolarRadiationLux(radiation: number): number | undefined {
+  if (!isNumber(radiation)) {
+    return undefined;
+  }
+
+  // Conversion factor from https://github.com/home-assistant/core/blob/dev/homeassistant/components/ambient_station/__init__.py
+  return radiation / 0.0079;
+}
+
+// Returns the current date if rainAmount is greater than 0. Otherwise
+// returns undefined.
+/**
+ *
+ * @param rainAmount The amount of rain in the last hour
+ * @returns The current time if the rain amount last hour was more than 0. Otherwise returns undefined.
```

Comment:
How does `undefined` get processed on the MQTT client/HA side? Does it retain the last valid valid, or will it go undefined there too? 

For example, if it's currently raining, I expect this value to continuously update. But after it stops, it should retain the timestamp of the last time the hourly rain was > 0. And on HA that would usually show as relative datetime of "2 days ago" or similar.

Reply:
I noticed this is a common return value, but in most cases it's an error condition. This sensor acts a little different in that regard

Reply:
Undefined values wind up not getting set on the sensor, see https://github.com/neilenns/ambientweather2mqtt/blob/8f37578a46ac6d7d70e72c3b3bf3d59bbd756f37/src/controllers/weatherDataController.ts#L80. I should add a unit test for this to verify, but I'm pretty sure what will happen is no value gets set on the sensor so it will just re-send whatever the last timestamp was.

Reply:
Great, yeah that makes sense if nothing new is published it should just retain the current/previous value.

Reply:
Ok no this doesn't work at all, unit test proves it's completely wrong. I'll have to do some mucking about to fix this.

Reply:
Bummer. Good catch though

Reply:
Ok I think this is fixed now.